### ISSUE TITLE
where is the definition of SAFE_HEAP_LOAD_i64_1_U_1

### ISSUE THREAD

**Issue by ssethwwang** [emscripten-core/emscripten#20892]
Sorry I cannot find the definition?

**Comment by sbc100** [comment]
All of the SAFE_HEAP_LOAD functions are defined in https://github.com/emscripten-core/emscripten/blob/main/src/runtime_safe_heap.js I believe.   I don't see one called `SAFE_HEAP_LOAD_i64_1_U_1` .. where is that reference coming from?   If its coming from emscripten itself can you share the link options that triggered it?

**Comment by kripken** [comment]
There are `SAFE_HEAP` methods defined in JS for JS use (as linked to in the previous comment), and others defined inside the wasm. I think `SAFE_HEAP_LOAD_i64_1_U_1` is inside the wasm. If so, it is generated by Binaryen's SafeHeap pass,

https://github.com/WebAssembly/binaryen/blob/main/src/passes/SafeHeap.cpp

**Comment by sbc100** [comment]
> There are `SAFE_HEAP` methods defined in JS for JS use (as linked to in the previous comment), and others defined inside the wasm. I think `SAFE_HEAP_LOAD_i64_1_U_1` is inside the wasm. If so, it is generated by Binaryen's SafeHeap pass,
> 
> https://github.com/WebAssembly/binaryen/blob/main/src/passes/SafeHeap.cpp

And does it also define those functions whenever it uses them?  It could they end up being imported?

**Comment by kripken** [comment]
Yes, it generates those functions as it generates calls to them. They are never imported. They do call imports like `segfault()` if they determine a fault has occurred.
